### ISSUE TITLE
Theos for the App Store

### ISSUE THREAD

**Issue by kirb** [theos/theos#286]
Theos should be able to build apps for releasing on the App Store. This would allow apps to be built on a non-Mac platform. Theos can already support building a full-blown real-deal app, including app extensions and an Apple Watch extension, without using anything that’d be considered private API. It should be completely possible to install a built app with a dev certificate on a connected phone, to debugging it with lldb (maybe even over wifi?), to uploading it to iTunes Connect to be pushed to TestFlight or the App Store. I don’t think there’s any reason Apple can reject apps built and submitted using Theos, if it all works properly.

### Prelude stuff:
* **Merge #278 so Swift works again.**
* **Release a working Swift compiler for building for Darwin/iOS on Linux, based on the official Swift.org builds for Linux.**  
   Probably also Cygwin but we’re discouraging use of Cygwin now.
* **Complete the unifying of the Darwin targets (#242).**  
   This would make it far less painful to create some App Store targets based on top of the existing `iphone`, `watchos`, etc targets and keep every platform and target in sync.
* **Build iOS/tvOS app and app extension templates.**  
   We already have application, and application_swift is coming in theos/templates#7. Also create tvos_application and tvos_application_swift, as well as templates for each kind of app extension that Theos can currently build without trouble (only in either ObjC or Swift, not both, to avoid clogging up the NIC list).

### Then, on a development branch:
* **Build a target based on the current `iphone` targets that set extra defaults for App Store distribution.**  
   Such as the app being staged to `/Payload` rather than `/Applications`, to enable Bitcode generation for release builds (required to submit a watchOS/tvOS app; not required but recommended for iOS apps), and possibly to disable codesigning on Theos’s part (see below).
* **Implement or find an existing cross-platform tool to generate .car asset catalogs.**  
   Seems like iOS 11 requires the app icon to be in a .car now? Either way it’d be cleaner to support this, both for App Store apps and any bundle built for a jailbreak package.
* **Figure out how we’ll support storyboards/xibs.**  
   While there’s some experimental storyboardc/nib compilers out there, an Interface Builder-like app would still be needed to create them. In a discussion we had a few weeks ago, we figured a primitive storyboard compiler would be fine to initially provide only Good Enough support for launch screens. Everything else can be achieved in code until a cross-platform interface builder app exists and is mature enough.
* **Add `ipa` packaging target.**  
   This would just zip up the staging directory and rename the file to end in .ipa.
* **Add `ipa` install target.**
   Most likely integrating with Impactor. Impactor asking for password all the time would be tedious and I’d have to ask saurik what can be done about that, but otherwise Impactor handles everything without user interaction. Also, does Impactor have a CLI mode so it can be used over SSH/on a build server?
* **Get iTunes Connect uploads working.**  
   Apple’s Application Loader.app is an interface to a CLI Java program called iTMSTransporter (and amusingly includes a Windows batch file for launching it). We could extract it from the Application Loader DMG and use it either directly or via fastlane’s [deliver](https://docs.fastlane.tools/actions/deliver/). We could try implementing our own equivalent in Perl or Go or something but I think we’d be far safer using Apple’s implementation especially as it clearly seems to be cross-platform.
* **Determine what other integration we should have with fastlane.**  
   Should we use more components of it just to take advantage of a more well-known codebase owned by someone who’s paid by Google to maintain it? (It should only be required when a project is targeting App Store, of course.)
* **Determine if we can make CocoaPods, Carthage, or Swift Package Manager work.**  
   Would be a major asshole (too unreliable) to parse stuff out of an .xcodeproj and try and make it build using Theos. Swift Package Manager probably won’t work either since as far as I’m aware it’ll want to build the package using `xcodebuild`, which expects an .xcodeproj.

-----

This is mostly just a jumble of thoughts I’ve wanted to get written down for a while; hopefully later on this will be more clarified.

**Comment by uroboro** [comment]
Regarding storyboards, here's a [python ibtool](https://github.com/davidquesada/ibtool). Potentially combine with this [py2perl](https://github.com/bwright/python-to-perl) project so we don't add another dependency.

**Comment by kabiroberai** [comment]
It's possible to use CocoaPods outside of Xcode with [Rome](https://github.com/CocoaPods/Rome), which builds dynamic frameworks. It's been a year since it was updated though so I don't know whether it still works.

**Comment by kirb** [comment]
@kabiroberai, that seems to just call `xcodebuild`, so we’re still stuck for non-macOS platforms.

**Comment by wstyres** [comment]
For Assets.car compiling, I've found that macOS (at least) contains `actool` and `PlistBuddy` that help with compiling a carfile for use in an app. 

Provided that you have already built an Assets.xcassets in Xcode, you can run `/Applications/Xcode.app/Contents/Developer/usr/bin/actool (location of  Assets.xcassets) --compile (location of where to place the images and Assets.car i.e. a resources directory) --platform iphoneos  --minimum-deployment-target (minimum deployment) --app-icon (name of image set for using an AppIcon, not necessary) --launch-image (name of image set for using LaunchImages, not necessary) --output-partial-info-plist tmp.plist` 

This will compile the Assets.car file and create a tmp.plist file which contains useful keys for implementing the AppIcon and LaunchImages in your app's Info.plist. Using PlistBuddy, we can use the following command `/usr/libexec/PlistBuddy -x -c "Merge tmp.plist" <Info.plist location>` to merge the temporary plist into the application's Info.plist which will then be compiled into the app in a later stage.

`rm tmp.plist` will remove the temporary plist. Although this isn't necessary at all because `actool` will overwrite the plist on next run, it frees up the git repo for me.

To summarize, my makefile looks a little like this (forgive my implementation, idk how to Makefile that well)

```
do::
	/Applications/Xcode.app/Contents/Developer/usr/bin/actool Assets.xcassets --compile ./Resources --platform iphoneos  --minimum-deployment-target 8.0 --app-icon AppIcon --launch-image LaunchImage --output-partial-info-plist tmp.plist
	/usr/libexec/PlistBuddy -x -c "Merge tmp.plist" ./Resources/Info.plist
	rm tmp.plist
```

**Comment by serjepatoff** [comment]
If you replicate all compiling and linking flags from Xcode, and also signing, packaging and submission in exactly the same manner, it'd possible to talk about "possibility to detect" instead of "reason to reject". :)

Thank you guys for this great project.

**Comment by kirb** [comment]
There definitely could be slight differences. One good example is that for a while the OSS release of libtapi simply didn’t work and nobody was really sure why. Clearly Apple forgot to include something, or intentionally excluded something, or we simply didn’t know how to build it correctly because of a lack of documentation. Another example, swift.org has a scary warning box saying only Xcode-packaged releases of Swift are allowed for App Store submission. So, what’s the difference between the swift.org toolchain and Xcode’s Swift toolchain, and why? Or is it just there as a disclaimer Apple can point to if your app breaks?

I don’t think there would be good reason to reject apps built with a completely-OSS toolchain, but yes it’s possible there are subtleties that could break assumptions made in the App Store processing, or later on if Apple needs to re-process the app due to an issue they discover in the processing. Lots of unknowns there and since Apple has to support the many thousands of apps on the store, the responsibility to fix Apple’s fix if it breaks the app lies on the developer and their toolchain vendor. (Apple could still break things like Mono Touch/Xamarin of course, but Apple would know to take care of their nuances since they’re in widespread use.)

**Comment by amin1995** [comment]
F
